content/en/apps/tutorials/application-tests.md

```diff
@@ -0,0 +1,600 @@
+
+Let's look at a more detailed example. <a name="assessment-form-test">Here</a> is a test case for the Assessment form that was covered in the [previous tutorial]({{< ref "apps/tutorials/app-forms" >}}):
+
+{{< highlight js "linenos=table,hl_lines=26 31 36,anchorlinenos=true,lineanchors=assessment-form-test" >}}
+  it('unit test confirming assessment with cough since 7 days', async () => {
+    // Load the assessment form and fill 'yes' on the first page and '7' on the second page
+    const result = await harness.fillForm('assessment', ['yes'], ['7']);
+
+    // Verify that the form successfully got submitted
+    expect(result.errors).to.be.empty;
+
+    // Verify some attributes on the resulting report
+    expect(result.report.fields).to.deep.include({
+      patient_name: 'Patient Name',
+      group_assessment: {
+        cough: 'yes',
+        cough_duration: '7',
+      }
+    });
+  });
+{{< /highlight >}}
+
+In [line 3](#assessment-form-test-3) above, the method [harness.fillForm()](https://docs.communityhealthtoolkit.org/cht-conf-test-harness/Harness.html#fillForm) fills the specified form with the given input (answers).
+
+Depending on the form design, the number of inputs to be filled can be large. The inputs are often repeated within a single test or across multiple tests with little or no variation. It is a good idea to keep them in a separate file and refer them from the tests as required. You can also introduce some variations in the inputs using function parameters.
+
+Example: `form-inputs.js`
+```js {linenos=table}
+module.exports = {
+  assessments: {
+    no_cough: [
+      ['no']
+    ],
+    cough: (days) => [
+      ['yes'],
+      [ days ]
+    ]
+  }
+};
+```
+
+In the test below, the inputs are used from the `form-inputs.js` file above:
+```js
+const { assessments } = require('../form-inputs');
+...
+const result = await harness.fillForm('assessment', ...assessments.cough('3'));
+```
+
+The test files are usually grouped in folders to read and run them easily. One way of grouping them is by creating folders for each of the components that can be tested: forms, contact summary, tasks, and targets.
+
+![alt_text](tests-dir.png "Test directory structure")
+
+### Testing Forms
+<table>
+  <tr>
+    <th colspan='2'> What do you test? </th>
+  </tr>
+  <tr>
+   <td>Minimum:</td>
+   <td>Filling a form with the most common options should not result in any error.</td>
+  </tr>
+  <tr>
+   <td>Ideal:</td>
+   <td>All input combinations and constraints are tested with the report fields.</td>
+  </tr>
+</table>
+
+[Previous example](#assessment-form-test) demonstrates test for the app form. More test cases can be added by changing the inputs.
+
+You can also test contact forms using test harness. To fill a contact form, use <code>[fillContactForm(contactType, …answers)](https://docs.communityhealthtoolkit.org/cht-conf-test-harness/Harness.html#fillContactForm)</code>.
+
+Example:
+```js {linenos=table}
```

Comment:
Do we need {linenos=table}? I am not sure but I noticed some code snippet has line numbers and some don't. Just a suggestion, may we can exclude from all others where we are not referencing line numbers of the code snippet just to make it consistent?

Reply:
Excluded 1`linenos` from everywhere where not referencing. Thanks!